client/core/types.go

```diff
@@ -126,41 +126,126 @@ type Match struct {
+	Refund        *Coin             `json:"refund,omitempty"`
 	Stamp         uint64            `json:"stamp"`
 	IsCancel      bool              `json:"isCancel"`
 }
 
-// matchFromMetaMatch constructs a *Match from an *db.MetaMatch
-func matchFromMetaMatch(metaMatch *db.MetaMatch) *Match {
-	userMatch, proof := metaMatch.Match, &metaMatch.MetaData.Proof
-	match := &Match{
-		MatchID:  userMatch.MatchID[:],
-		Status:   userMatch.Status,
-		Revoked:  proof.IsRevoked(),
-		Rate:     userMatch.Rate,
-		Qty:      userMatch.Quantity,
-		Side:     userMatch.Side,
-		FeeRate:  userMatch.FeeRateSwap,
-		Refund:   []byte(proof.RefundCoin),
-		Stamp:    metaMatch.MetaData.Stamp,
-		IsCancel: userMatch.Address == "",
+// A coin encodes both the coin ID and the asset-dependent string representation
+// of the coin ID.
+type Coin struct {
+	ID       dex.Bytes `json:"id"`
+	StringID string    `json:"stringID"`
+	AssetID  uint32    `json:"assetID"`
+	Symbol   string    `json:"symbol"`
+	// Confs is populated only if this is a swap coin and we are waiting for
+	// confirmations e.g. when this is the maker's swap and we're in
+	// MakerSwapCast, or this is the takers swap, and we're in TakerSwapCast.
+	Confs *Confirmations `json:"confs,omitempty"`
+}
+
+// NewCoin constructs a new Coin.
+func NewCoin(assetID uint32, coinID []byte) *Coin {
+	return &Coin{
+		ID:       coinID,
+		StringID: coinIDString(assetID, coinID),
+		AssetID:  assetID,
+		Symbol:   unbip(assetID),
 	}
+}
+
+// Confirmations is the confirmation count and confirmation requirement of a
+// SwapCoin.
+type Confirmations struct {
+	Count    int64 `json:"count"`
+	Required int64 `json:"required"`
+}
 
-	if userMatch.Side == order.Maker {
-		match.Swap = []byte(proof.MakerSwap)
-		match.CounterSwap = []byte(proof.TakerSwap)
-		match.Redeem = []byte(proof.MakerRedeem)
-		match.CounterRedeem = []byte(proof.TakerRedeem)
-	} else {
-		match.Swap = []byte(proof.TakerSwap)
-		match.CounterSwap = []byte(proof.MakerSwap)
-		match.Redeem = []byte(proof.TakerRedeem)
-		match.CounterRedeem = []byte(proof.MakerRedeem)
+// SetConfirmations sets the Confs field of the Coin.
+func (c *Coin) SetConfirmations(confs, confReq int64) {
+	c.Confs = &Confirmations{
+		Count:    confs,
+		Required: confReq,
+	}
+}
+
+// matchFromMetaMatch constructs a *Match from an Order and a *MetaMatch.
+// This function is intended for use with inactive matches. For active matches,
+// use matchFromMetaMatchWithConfs.
+func matchFromMetaMatch(ord order.Order, metaMatch *db.MetaMatch) *Match {
+	return matchFromMetaMatchWithConfs(ord, metaMatch, 0, 0, 0, 0)
+}
+
+// matchFromMmatchFromMetaMatchWithConfs constructs a *Match from a *MetaMatch,
+// and sets the confirmations for swaps-in-waiting.
+func matchFromMetaMatchWithConfs(ord order.Order, metaMatch *db.MetaMatch, swapConfs, swapReq, counterSwapConfs, counterReq int64) *Match {
+	side := metaMatch.Match.Side
+	sell := ord.Trade().Sell
```

Comment:
`ord.Trade()` is nil for cancel orders, and thus panic, and while this shouldn't be called for a cancel order match (I think) I'd feel better if this function had logic to deal with that possibility.